File: src/block/build_ghost.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
! subroutine build_ghost: Build all information for ghost cells by 
! mirroring information from interior cells.  The action includes:
!   1. define indices and build connectivities for ghost nodes, faces, 
!      and cells.  In the same loop, mirror the coordinates of interior 
!      nodes to ghost nodes.
!   2. compute center coordinates for faces for ghost cells.
!   3. compute normal vectors and areas for faces for ghost cells.
!   4. compute center coordinates for ghost cells.
!   5. compute volume for ghost cells.
! NOTE: all the metric, type and connnectivities data passed in this 
!   subroutine are SHARED arrays rather than interior arrays.  The 
!   indices for ghost information should be carefully treated.  All the 
!   ghost indices are negative in shared arrays.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
subroutine build_ghost(msh, bndfcs, fctpn, cltpn, clgrp, &
    fcnds, fccls, clnds, clfcs, &
    ndcrd, fccncrd, fcnormal, fcarea, clcncrd, clvol)
implicit none
include 'mesh.inc'
! meta data.
type(mesh), intent(in) :: msh
integer(4), intent(in) :: bndfcs(0:1, 0:msh%ngstcell-1)
integer(4), intent(inout) :: fctpn(-msh%ngstface:msh%nface-1)
integer(4), intent(inout) :: cltpn(-msh%ngstcell:msh%ncell-1)
integer(4), intent(inout) :: clgrp(-msh%ngstcell:msh%ncell-1)
! connectivity.
integer(4), intent(inout) :: fcnds(0:msh%fcmnd, -msh%ngstface:msh%nface-1)
integer(4), intent(inout) :: fccls(0:3, -msh%ngstface:msh%nface-1)
integer(4), intent(inout) :: clnds(0:msh%clmnd, -msh%ngstcell:msh%ncell-1)
integer(4), intent(inout) :: clfcs(0:msh%clmfc, -msh%ngstcell:msh%ncell-1)
! geometry.
real(FPKIND), intent(inout) :: ndcrd(0:msh%ndim-1, -msh%ngstnode:msh%nnode-1)
real(FPKIND), intent(inout) :: fccncrd(0:msh%ndim-1, -msh%ngstface:msh%nface-1)
real(FPKIND), intent(inout) :: fcnormal(0:msh%ndim-1, -msh%ngstface:msh%nface-1)
real(FPKIND), intent(inout) :: fcarea(-msh%ngstface:msh%nface-1)
real(FPKIND), intent(inout) :: clcncrd(0:msh%ndim-1,-msh%ngstcell:msh%ncell-1)
real(FPKIND), intent(inout) :: clvol(-msh%ngstcell:msh%ncell-1)

integer(4), allocatable :: gstndmap(:)

integer(4) :: nnd, nfc
real(FPKIND) :: ndist, vol
real(FPKIND) :: lvec(0:msh%ndim-1), radvec(0:msh%ndim-1, 0:msh%fcmnd-1)

integer(4) :: fcnnd, clnnd, clnfc
integer(4) :: igcl, igfc, ignd
integer(4) :: iicl, ibfc
integer(4) :: ifc, ind, icl
integer(4) :: it, itt

integer(4) :: mk_found

allocate(gstndmap(0:msh%nnode-1))
gstndmap(:) = msh%nnode ! set to the least impossible value.

! create ghost entities and build connectivities and by the way mirror node 
! coordinate.
ignd = -1
igfc = -1
do igcl = -1, -msh%ngstcell, -1
    ibfc = bndfcs(0,-igcl-1)    ! indicate current boundary face.
    iicl = fccls(0,ibfc)    ! indicate current interior cell.
    fcnnd = fcnds(0,ibfc)   ! number of nodes in current boundary face.
    clnnd = clnds(0,iicl)   ! number of nodes in current interior cell.
    clnfc = clfcs(0,iicl)   ! number of faces in current interior cell.
    ! copy cell type.
    cltpn(igcl) = cltpn(iicl)
    ! copy cell group.
    clgrp(igcl) = clgrp(iicl)
    ! process node list in ghost cell.
    clnds(:,igcl) = clnds(:,iicl)   ! copy nodes from current interior cell.
    loop_clnnd: do it = 1, clnnd
        ind = clnds(it,iicl)
        ! try to find the node in boundary face.
        mk_found = 0
        do itt = 1, fcnnd
            if (ind .eq. fcnds(itt,ibfc)) then
                mk_found = 1
                exit
            end if
        end do
        ! if it's not in boundary face, it should be a ghost node.
        if (mk_found .eq. 0) then
            gstndmap(ind) = ignd    ! record map for faces processing.
            clnds(it,igcl) = ignd   ! save to clnds.
            ! mirror coordinate of ghost cell.
            ! NOTE: fcnormal always point outward.
            lvec(:) = fccncrd(:,ibfc) - ndcrd(:,ind)
            ndist = dot_product(lvec(:), fcnormal(:,ibfc))
            ndcrd(:,ignd) = ndcrd(:,ind) + 2*ndist*fcnormal(:,ibfc)
            ! decrement ghost node counter.
            ignd = ignd - 1
        end if
    end do loop_clnnd
    ! set the relating cell as ghost cell.
    fccls(1,ibfc) = igcl
    ! process face list in ghost cell.
    clfcs(:,igcl) = clfcs(:,iicl)   ! copy faces in current interior cell.
    loop_clnfc: do it = 1, clnfc
        ifc = clfcs(it,igcl)        ! indicate face to be processed.
        if (ifc .eq. ibfc) cycle    ! if boundary face then skip.
        fctpn(igfc) = fctpn(ifc)    ! copy face type.
        fccls(0,igfc) = igcl        ! save to fccls.
        clfcs(it,igcl) = igfc       ! save to clfcs.
        ! face-to-node connectivity.
        fcnds(:,igfc) = fcnds(:,ifc)    ! copy from interior face.
        do itt = 1, fcnds(0,igfc)
            ind = fcnds(itt,igfc)
            if (gstndmap(ind) .eq. msh%nnode) cycle ! no map data: on bnd face.
            fcnds(itt,igfc) = gstndmap(ind)     ! save to fcnds.
        end do
        ! decrement ghost face counter.
        igfc = igfc - 1
    end do loop_clnfc
    ! erase node map record.
    do it = 1, clnnd
        gstndmap(clnds(it,iicl)) = msh%nnode
    end do
end do
! tests.
!if (ignd .ne. -ngstnode-1) print*, "trouble about node:", ignd, ngstnode
!if (igfc .ne. -ngstface-1) print*, "trouble about face:", igfc, ngstface
!do it = 0, nnode-1
!    if (gstndmap(it) .ne. nnode) print*, "trouble about map:", it, gstndmap(it)
!end do

! To this place, all ghost connectivities are computed.
! Start compute ghost metrics.

! compute ghost face center coordinate.
fccncrd(:,-msh%ngstface:-1) = 0.d0
do ifc = -1, -msh%ngstface, -1  ! NOTE: indicate ghost face.
    nnd = fcnds(0,ifc)
    do it = 1, nnd
        ind = fcnds(it,ifc)
        fccncrd(:,ifc) = fccncrd(:,ifc) + ndcrd(:,ind)
    end do
    fccncrd(:,ifc) = fccncrd(:,ifc) / nnd
end do

! compute face normal vector and face area.
fcnormal(:,-msh%ngstface:-1) = 0.d0
if (msh%ndim .eq. 2) then
    do ifc = -1, -msh%ngstface, -1  ! NOTE: indicate ghost face.
        nnd = 2 ! face must be a line.
        lvec(:) = ndcrd(:,fcnds(2,ifc)) - ndcrd(:,fcnds(1,ifc))
        fcnormal(0,ifc) = lvec(1)
        fcnormal(1,ifc) = -lvec(0)
        ! compute face area.
        fcarea(ifc) = sqrt(sum(fcnormal(:,ifc)**2))
        ! make face normal unitized.
        fcnormal(:,ifc) = fcnormal(:,ifc)/fcarea(ifc)
    end do
else if (msh%ndim .eq. 3) then
    do ifc = -1, -msh%ngstface, -1  ! NOTE: indicate ghost face.
        nnd = fcnds(0,ifc)
        ! compute radial vector.
        do it = 0, nnd-1    ! NOTE: iterator starts from 0.
            ind = fcnds(it+1,ifc)
            radvec(:,it) = ndcrd(:,ind) - fccncrd(:,ifc)
        end do
        ! compute cross product, summation over all subtriangle.
        ! NOTE: this requires specific ordering in face node definition.
        fcnormal(0,ifc) = &
            radvec(1,nnd-1)*radvec(2,0) - radvec(2,nnd-1)*radvec(1,0)
        fcnormal(1,ifc) = &
            radvec(2,nnd-1)*radvec(0,0) - radvec(0,nnd-1)*radvec(2,0)
        fcnormal(2,ifc) = &
            radvec(0,nnd-1)*radvec(1,0) - radvec(1,nnd-1)*radvec(0,0)
        do it = 1, nnd-1
            fcnormal(0,ifc) = fcnormal(0,ifc) + &
                radvec(1,it-1)*radvec(2,it) - radvec(2,it-1)*radvec(1,it)
            fcnormal(1,ifc) = fcnormal(1,ifc) + &
                radvec(2,it-1)*radvec(0,it) - radvec(0,it-1)*radvec(2,it)
            fcnormal(2,ifc) = fcnormal(2,ifc) + &
                radvec(0,it-1)*radvec(1,it) - radvec(1,it-1)*radvec(0,it)
        end do
        ! compute face area*2.
        fcarea(ifc) = sqrt(sum(fcnormal(:,ifc)**2))
        ! make face normal unitized.
        fcnormal(:,ifc) = fcnormal(:,ifc)/fcarea(ifc)
        ! get face area.
        fcarea(ifc) = fcarea(ifc)/2.d0
    end do
end if

! compute center point coordinate for each ghost cell.
clcncrd(:,-msh%ngstcell:-1) = 0.d0
do icl = -1, -msh%ngstcell, -1
    nnd = clnds(0,icl)
    do it = 1, nnd
        ind = clnds(it, icl)
        clcncrd(:,icl) = clcncrd(:,icl) + ndcrd(:,ind)
    end do
    clcncrd(:,icl) = clcncrd(:,icl)/nnd
end do

! compute volume for each ghost cell.
clvol(-msh%ngstcell:-1) = 0.d0
do icl = -1, -msh%ngstcell, -1
    nfc = clfcs(0,icl)
    do it = 1, nfc
        ifc = clfcs(it, icl)
        ! calculate volume associated with each bounding face.
        lvec(:) = fccncrd(:,ifc) - clcncrd(:,icl)
        vol = dot_product(lvec(:), fcnormal(:,ifc)) * fcarea(ifc)
        ! check if the normal vector for ghost face point outward.
        if (vol .lt. 0.d0) then
            if (ifc .lt. 0) then
                fcnormal(:,ifc) = -fcnormal(:,ifc)
            end if
            vol = -vol
        end if
        ! accumulate the volume for icell.
        clvol(icl) = clvol(icl) + vol
    end do
    clvol(icl) = clvol(icl)/msh%ndim
end do

deallocate(gstndmap)

end subroutine build_ghost
! vim:set nu et tw=80 ts=4 sw=4 cino=>4:

File: src/blockc/build_ghost.c
/*
 * Copyright (C) 2008-2010 Yung-Yu Chen.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "solvcon.h"
/*
 * subroutine build_ghost: Build all information for ghost cells by 
 * mirroring information from interior cells.  The action includes:
 *
 * 1. define indices and build connectivities for ghost nodes, faces, 
 *    and cells.  In the same loop, mirror the coordinates of interior 
 *    nodes to ghost nodes.
 * 2. compute center coordinates for faces for ghost cells.
 * 3. compute normal vectors and areas for faces for ghost cells.
 * 4. compute center coordinates for ghost cells.
 * 5. compute volume for ghost cells.
 *
 * NOTE: all the metric, type and connnectivities data passed in this 
 * subroutine are SHARED arrays rather than interior arrays.  The 
 * indices for ghost information should be carefully treated.  All the 
 * ghost indices are negative in shared arrays.
 */
int build_ghost(MeshData *msd, int *bndfcs) {
    int nnd, nfc;
    // pointers.
    int *pbndfcs;
    int *pfctpn, *pcltpn, *pclgrp;
    int *pgfctpn, *pgcltpn, *pgclgrp;
    int *pfcnds, *pfccls, *pclnds, *pclfcs;
    int *pgfcnds, *pgfccls, *pgclnds, *pgclfcs;
    FPTYPE *pndcrd, *p2ndcrd, *pgndcrd;
    FPTYPE *pfccnd, *pfcnml, *pfcara, *pclcnd, *pclvol;
    // buffers.
    int *gstndmap;
    // scalars.
    int mk_found;
    FPTYPE vol, vob, voc;
    FPTYPE du0, du1, du2, dv0, dv1, dv2, dw0, dw1, dw2;
    // arrays.
    FPTYPE cfd[FCMND+2][msd->ndim];
    FPTYPE crd[msd->ndim];
    FPTYPE radvec[FCMND][msd->ndim];
    // iterators.
    int ind, ibfc, icl, inl, inf, idm, ifl, ifc;
    int ignd, igfc, igcl;
    int it;

    gstndmap = (int *)malloc((size_t)msd->nnode*sizeof(int));
    for (ind=0; ind<msd->nnode; ind++) {
        gstndmap[ind] = msd->nnode; // initialize to the least possible value.
    };

    // create ghost entities and buil connectivities and by the way mirror node
    // coordinate.
    ignd = -1;
    igfc = -1;
    pbndfcs = bndfcs;
    pgndcrd = msd->ndcrd - msd->ndim;
    pgfctpn = msd->fctpn - 1;
    pgfcnds = msd->fcnds - (FCMND+1);
    pgfccls = msd->fccls - FCREL;
    pgcltpn = msd->cltpn - 1;
    pgclgrp = msd->clgrp - 1;
    pgclnds = msd->clnds - (CLMND+1);
    pgclfcs = msd->clfcs - (CLMFC+1);
    for (igcl=-1; igcl>=-msd->ngstcell; igcl--) {
        ibfc = pbndfcs[0];
        pfctpn = msd->fctpn + ibfc;
        pfcnds = msd->fcnds + ibfc*(FCMND+1);
        pfccls = msd->fccls + ibfc*FCREL;
        pfccnd = msd->fccnd + ibfc*msd->ndim;
        pfcnml = msd->fcnml + ibfc*msd->ndim;
        icl = pfccls[0];
        pcltpn = msd->cltpn + icl;
        pclgrp = msd->clgrp + icl;
        pclnds = msd->clnds + icl*(CLMND+1);
        pclfcs = msd->clfcs + icl*(CLMFC+1);
        // copy cell type and group.
        pgcltpn[0] = pcltpn[0];
        pgclgrp[0] = pclgrp[0];
        // process node list in ghost cell.
        for (inl=0; inl<=CLMND; inl++) {    // copy nodes from current in-cell.
            pgclnds[inl] = pclnds[inl];
        };
        for (inl=1; inl<=pclnds[0]; inl++) {
            ind = pclnds[inl];
            pndcrd = msd->ndcrd + ind*msd->ndim;
            // try to find the node in the boundary face.
            mk_found = 0;
            for (inf=1; inf<=pfcnds[0]; inf++) {
                if (ind == pfcnds[inf]) {
                    mk_found = 1;
                    break;
                };
            };
            // if not found, it should be a ghost node.
            if (mk_found == 0) {
                gstndmap[ind] = ignd;   // record map for face processing.
                pgclnds[inl] = ignd;    // save to clnds.
                // mirror coordinate of ghost cell.
                // NOTE: fcnml always points outward.
                vol = 0.0;
                for (idm=0; idm<msd->ndim; idm++) {
                    vol += (pfccnd[idm] - pndcrd[idm]) * pfcnml[idm];
                };
                for (idm=0; idm<msd->ndim; idm++) {
                    pgndcrd[idm] = pndcrd[idm] + 2*vol*pfcnml[idm];
                };
                // decrement ghost node counter.
                ignd -= 1;
                pgndcrd -= msd->ndim;
            };
        };
        // set the relating cell as ghost cell.
        pfccls[1] = igcl;
        // process face list in ghost cell.
        for (ifl=0; ifl<=CLMFC; ifl++) {
            pgclfcs[ifl] = pclfcs[ifl]; // copy in-face to ghost.
        };
        for (ifl=1; ifl<=pclfcs[0]; ifl++) {
            ifc = pclfcs[ifl];  // the face to be processed.
            if (ifc == ibfc) continue;  // if boundary face then skip.
            pfctpn = msd->fctpn + ifc;
            pfcnds = msd->fcnds + ifc*(FCMND+1);
            pgfctpn[0] = pfctpn[0]; // copy face type.
            pgfccls[0] = igcl;  // save to ghost fccls.
            pgclfcs[ifl] = igfc;    // save to ghost clfcs.
            // face-to-node connectivity.
            for (inf=0; inf<=FCMND; inf++) {
                pgfcnds[inf] = pfcnds[inf];
            };
            for (inf=1; inf<=pgfcnds[0]; inf++) {
                ind = pgfcnds[inf];
                if (gstndmap[ind] != msd->nnode) {
                    pgfcnds[inf] = gstndmap[ind];   // save gstnode to fcnds.
                };
            };
            // decrement ghost face counter.
            igfc -= 1;
            pgfctpn -= 1;
            pgfcnds -= FCMND+1;
            pgfccls -= FCREL;
        };
        // erase node map record.
        for (inl=1; inl<=pclnds[0]; inl++) {
                gstndmap[pclnds[inl]] = msd->nnode;
        };
        // advance pointers.
        pbndfcs += 2;
        pgcltpn -= 1;
        pgclgrp -= 1;
        pgclnds -= CLMND+1;
        pgclfcs -= CLMFC+1;
    };
    free(gstndmap);

    // compute ghost face centroids.
    pfcnds = msd->fcnds - (FCMND+1);
    pfccnd = msd->fccnd - msd->ndim;
    if (msd->ndim == 2) {
        // 2D faces must be edge.
        for (ifc=-1; ifc>=-msd->ngstface; ifc--) {
            // point 1.
            ind = pfcnds[1];
            pndcrd = msd->ndcrd + ind*msd->ndim;
            pfccnd[0] = pndcrd[0];
            pfccnd[1] = pndcrd[1];
            // point 2.
            ind = pfcnds[2];
            pndcrd = msd->ndcrd + ind*msd->ndim;
            pfccnd[0] += pndcrd[0];
            pfccnd[1] += pndcrd[1];
            // average.
            pfccnd[0] /= 2;
            pfccnd[1] /= 2;
            // advance pointers.
            pfcnds -= FCMND+1;
            pfccnd -= msd->ndim;
        };
    } else if (msd->ndim == 3) {
        for (ifc=-1; ifc>=-msd->ngstface; ifc--) {
            // find averaged point.
            cfd[0][0] = cfd[0][1] = cfd[0][2] = 0.0;
            nnd = pfcnds[0];
            for (inf=1; inf<=nnd; inf++) {
                ind = pfcnds[inf];
                pndcrd = msd->ndcrd + ind*msd->ndim;
                cfd[inf][0]  = pndcrd[0];
                cfd[0  ][0] += pndcrd[0];
                cfd[inf][1]  = pndcrd[1];
                cfd[0  ][1] += pndcrd[1];
                cfd[inf][2]  = pndcrd[2];
                cfd[0  ][2] += pndcrd[2];
            };
            cfd[nnd+1][0] = cfd[1][0];
            cfd[nnd+1][1] = cfd[1][1];
            cfd[nnd+1][2] = cfd[1][2];
            cfd[0][0] /= nnd;
            cfd[0][1] /= nnd;
            cfd[0][2] /= nnd;
            // calculate area.
            pfccnd[0] = pfccnd[1] = pfccnd[2] = voc = 0.0;
            for (inf=1; inf<=nnd; inf++) {
                crd[0] = (cfd[0][0] + cfd[inf][0] + cfd[inf+1][0])/3;
                crd[1] = (cfd[0][1] + cfd[inf][1] + cfd[inf+1][1])/3;
                crd[2] = (cfd[0][2] + cfd[inf][2] + cfd[inf+1][2])/3;
                du0 = cfd[inf][0] - cfd[0][0];
                du1 = cfd[inf][1] - cfd[0][1];
                du2 = cfd[inf][2] - cfd[0][2];
                dv0 = cfd[inf+1][0] - cfd[0][0];
                dv1 = cfd[inf+1][1] - cfd[0][1];
                dv2 = cfd[inf+1][2] - cfd[0][2];
                dw0 = du1*dv2 - du2*dv1;
                dw1 = du2*dv0 - du0*dv2;
                dw2 = du0*dv1 - du1*dv0;
                vob = sqrt(dw0*dw0 + dw1*dw1 + dw2*dw2);
                pfccnd[0] += crd[0] * vob;
                pfccnd[1] += crd[1] * vob;
                pfccnd[2] += crd[2] * vob;
                voc += vob;
            };
            pfccnd[0] /= voc;
            pfccnd[1] /= voc;
            pfccnd[2] /= voc;
            // advance pointers.
            pfcnds -= FCMND+1;
            pfccnd -= msd->ndim;
        };
    };

    // compute ghost face normal vector and area.
    pfcnds = msd->fcnds - (FCMND+1);
    pfccnd = msd->fccnd - msd->ndim;
    pfcnml = msd->fcnml - msd->ndim;
    pfcara = msd->fcara - 1;
    if (msd->ndim == 2) {
        for (ifc=-1; ifc>=-msd->ngstface; ifc--) {
            // 2D faces are always lines.
            pndcrd = msd->ndcrd + pfcnds[1]*msd->ndim;
            p2ndcrd = msd->ndcrd + pfcnds[2]*msd->ndim;
            // face normal.
            pfcnml[0] = p2ndcrd[1] - pndcrd[1];
            pfcnml[1] = -(p2ndcrd[0] - pndcrd[0]);
            // face ara.
            pfcara[0] = sqrt(pfcnml[0]*pfcnml[0] + pfcnml[1]*pfcnml[1]);
            // normalize face normal.
            pfcnml[0] /= pfcara[0];
            pfcnml[1] /= pfcara[0];
            // advance pointers.
            pfcnds -= FCMND+1;
            pfcnml -= msd->ndim;
            pfcara -= 1;
        };
    } else if (msd->ndim == 3) {
        for (ifc=-1; ifc>=-msd->ngstface; ifc--) {
            // compute radial vector.
            nnd = pfcnds[0];
            for (inf=0; inf<nnd; inf++) {
                ind = pfcnds[inf+1];
                pndcrd = msd->ndcrd + ind*msd->ndim;
                radvec[inf][0] = pndcrd[0] - pfccnd[0];
                radvec[inf][1] = pndcrd[1] - pfccnd[1];
                radvec[inf][2] = pndcrd[2] - pfccnd[2];
            };
            // compute cross product.
            pfcnml[0] = radvec[nnd-1][1]*radvec[0][2]
                      - radvec[nnd-1][2]*radvec[0][1];
            pfcnml[1] = radvec[nnd-1][2]*radvec[0][0]
                      - radvec[nnd-1][0]*radvec[0][2];
            pfcnml[2] = radvec[nnd-1][0]*radvec[0][1]
                      - radvec[nnd-1][1]*radvec[0][0];
            for (ind=1; ind<nnd; ind++) {
                pfcnml[0] += radvec[ind-1][1]*radvec[ind][2]
                           - radvec[ind-1][2]*radvec[ind][1];
                pfcnml[1] += radvec[ind-1][2]*radvec[ind][0]
                           - radvec[ind-1][0]*radvec[ind][2];
                pfcnml[2] += radvec[ind-1][0]*radvec[ind][1]
                           - radvec[ind-1][1]*radvec[ind][0];
            };
            // compute face area.
            pfcara[0] = sqrt(pfcnml[0]*pfcnml[0] + pfcnml[1]*pfcnml[1]
                           + pfcnml[2]*pfcnml[2]);
            // normalize normal vector.
            pfcnml[0] /= pfcara[0];
            pfcnml[1] /= pfcara[0];
            pfcnml[2] /= pfcara[0];
            // get real face area.
            pfcara[0] /= 2.0;
            // advance pointers.
            pfcnds -= FCMND+1;
            pfccnd -= msd->ndim;
            pfcnml -= msd->ndim;
            pfcara -= 1;
        };
    };

    // compute cell centroids.
    pclnds = msd->clnds - (CLMND+1);
    pclfcs = msd->clfcs - (CLMFC+1);
    pclcnd = msd->clcnd - msd->ndim;
    if (msd->ndim == 2) {
        for (icl=-1; icl>=-msd->ngstcell; icl--) {
            // averaged point.
            crd[0] = crd[1] = 0.0;
            nnd = pclnds[0];
            for (inl=1; inl<=nnd; inl++) {
                ind = pclnds[inl];
                pndcrd = msd->ndcrd + ind*msd->ndim;
                crd[0] += pndcrd[0];
                crd[1] += pndcrd[1];
            };
            crd[0] /= nnd;
            crd[1] /= nnd;
            // weight centroid.
            pclcnd[0] = pclcnd[1] = voc = 0.0;
            nfc = pclfcs[0];
            for (ifl=1; ifl<=nfc; ifl++) {
                ifc = pclfcs[ifl];
                pfccnd = msd->fccnd + ifc*msd->ndim;
                pfcnml = msd->fcnml + ifc*msd->ndim;
                pfcara = msd->fcara + ifc;
                du0 = crd[0] - pfccnd[0];
                du1 = crd[1] - pfccnd[1];
                vob = fabs(du0*pfcnml[0] + du1*pfcnml[1]) * pfcara[0];
                voc += vob;
                dv0 = pfccnd[0] + du0/3;
                dv1 = pfccnd[1] + du1/3;
                pclcnd[0] += dv0 * vob;
                pclcnd[1] += dv1 * vob;
            };
            pclcnd[0] /= voc;
            pclcnd[1] /= voc;
            // advance pointers.
            pclnds -= CLMND+1;
            pclfcs -= CLMFC+1;
            pclcnd -= msd->ndim;
        };
    } else if (msd->ndim == 3) {
        for (icl=-1; icl>=-msd->ngstcell; icl--) {
            // averaged point.
            crd[0] = crd[1] = crd[2] = 0.0;
            nnd = pclnds[0];
            for (inl=1; inl<=nnd; inl++) {
                ind = pclnds[inl];
                pndcrd = msd->ndcrd + ind*msd->ndim;
                crd[0] += pndcrd[0];
                crd[1] += pndcrd[1];
                crd[2] += pndcrd[2];
            };
            crd[0] /= nnd;
            crd[1] /= nnd;
            crd[2] /= nnd;
            // weight centroid.
            pclcnd[0] = pclcnd[1] = pclcnd[2] = voc = 0.0;
            nfc = pclfcs[0];
            for (ifl=1; ifl<=nfc; ifl++) {
                ifc = pclfcs[ifl];
                pfccnd = msd->fccnd + ifc*msd->ndim;
                pfcnml = msd->fcnml + ifc*msd->ndim;
                pfcara = msd->fcara + ifc;
                du0 = crd[0] - pfccnd[0];
                du1 = crd[1] - pfccnd[1];
                du2 = crd[2] - pfccnd[2];
                vob = fabs(du0*pfcnml[0] + du1*pfcnml[1] + du2*pfcnml[2])
                    * pfcara[0];
                voc += vob;
                dv0 = pfccnd[0] + du0/4;
                dv1 = pfccnd[1] + du1/4;
                dv2 = pfccnd[2] + du2/4;
                pclcnd[0] += dv0 * vob;
                pclcnd[1] += dv1 * vob;
                pclcnd[2] += dv2 * vob;
            };
            pclcnd[0] /= voc;
            pclcnd[1] /= voc;
            pclcnd[2] /= voc;
            // advance pointers.
            pclnds -= CLMND+1;
            pclfcs -= CLMFC+1;
            pclcnd -= msd->ndim;
        };
    };

    // compute volume for each ghost cell.
    pclfcs = msd->clfcs - (CLMFC+1);
    pclcnd = msd->clcnd - msd->ndim;
    pclvol = msd->clvol - 1;
    for (icl=-1; icl>=-msd->ngstcell; icl--) {
        pclvol[0] = 0.0;
        for (it=1; it<=pclfcs[0]; it++) {
            ifc = pclfcs[it];
            pfccls = msd->fccls + ifc*FCREL;
            pfcnds = msd->fcnds + ifc*(FCMND+1);
            pfccnd = msd->fccnd + ifc*msd->ndim;
            pfcnml = msd->fcnml + ifc*msd->ndim;
            pfcara = msd->fcara + ifc;
            // calculate volume associated with each face.
            vol = 0.0;
            for (idm=0; idm<msd->ndim; idm++) {
                vol += (pfccnd[idm] - pclcnd[idm]) * pfcnml[idm];
            };
            vol *= pfcara[0];
            // check if need to reorder node definition and connecting cell
            // list for the face.
            if (vol < 0.0) {
                if (pfccls[0] == icl) {
                    for (idm=0; idm<msd->ndim; idm++) {
                        pfcnml[idm] = -pfcnml[idm];
                    };
                };
                vol = -vol;
            };
            // accumulate the volume for the cell.
            pclvol[0] += vol;
        };
        // calculate the real volume.
        pclvol[0] /= msd->ndim;
        // advance pointers.
        pclfcs -= CLMFC+1;
        pclcnd -= msd->ndim;
        pclvol -= 1;
    };

    return 0;
};
// vim: set ts=4 et:
